goal_src/jak2/engine/gfx/sprite/sprite.gc

```diff
@@ -5,5 +5,1073 @@
+    (set! (-> v0-0 vec-data) (-> v0-0 data))
+    (set! (-> v0-0 adgif-data) (the-as (inline-array adgif-shader) (&-> v0-0 data vec-data-size)))
+    v0-0
+    )
+  )
+
+(defmethod new sprite-array-3d ((allocation symbol) (type-to-make type) (group-0-size int) (group-1-size int))
+  "Allocate a sprite-array for 3d sprites.  There are two groups, each can contain the given number of sprites.
+   Group 1 size is zero in practice for 3d."
+  (let* ((sprite-count (+ group-0-size group-1-size))
+         (vec-data-size (* 3 sprite-count))
+         (adgif-data-size (* 5 sprite-count))
+         (v0-0 (object-new
+                 allocation
+                 type-to-make
+                 (the-as int (+ (-> type-to-make size) (* (+ adgif-data-size -1 vec-data-size) 16)))
+                 )
+               )
+         )
+    (set! (-> v0-0 num-sprites 0) group-0-size)
+    (set! (-> v0-0 num-sprites 1) group-1-size)
+    (set! (-> v0-0 num-valid 0) 0)
+    (set! (-> v0-0 num-valid 1) 0)
+    (set! (-> v0-0 vec-data) (-> v0-0 data))
+    (set! (-> v0-0 adgif-data) (the-as (inline-array adgif-shader) (&-> v0-0 data vec-data-size)))
+    v0-0
+    )
+  )
+
+(define *sprite-array-2d* (new 'global 'sprite-array-2d 1920 128))
+;; note that there are 0 group 1 3d's
+(define *sprite-array-3d* (new 'global 'sprite-array-3d 256 0))
+
+(defun sprite-set-3d-quaternion! ((sprite sprite-vec-data-3d) (quat quaternion))
+  "Set the quaternion for a 3d sprite.
+   NOTE: the w component of the quaternion is not stored and
+   the original w value is preserved."
+  (local-vars (v1-0 float) (v1-1 float))
+  (rlet ((vf0 :class vf)
+         (vf1 :class vf)
+         (vf2 :class vf)
+         )
+    (init-vf0-vector)
+    (cond
+      ((< (-> quat w) 0.0)
+       (.lvf vf1 (&-> sprite qx-qy-qz-sy quad))
+       (.lvf vf2 (&-> quat vec quad))
+       (.sub.vf vf1 vf0 vf2 :mask #b111)
+       (.svf (&-> sprite qx-qy-qz-sy quad) vf1)
+       (.mov v1-0 vf1)
+       )
+      (else
+        (.lvf vf1 (&-> sprite qx-qy-qz-sy quad))
+        (.lvf vf2 (&-> quat vec quad))
+        (.add.vf vf1 vf0 vf2 :mask #b111)
+        (.svf (&-> sprite qx-qy-qz-sy quad) vf1)
+        (.mov v1-1 vf1)
+        )
+      )
+    quat
+    )
+  )
+
+(defun sprite-get-3d-quaternion! ((out quaternion) (sprite sprite-vec-data-3d))
+  "Get a quaternion. It's stored with only 3 components, so qw is
+   recalculated."
+  (let ((f0-0 (-> sprite qx-qy-qz-sy x))
+        (f1-0 (-> sprite qx-qy-qz-sy y))
+        (f2-0 (-> sprite qx-qy-qz-sy z))
+        )
+    (set! (-> out x) f0-0)
+    (set! (-> out y) f1-0)
+    (set! (-> out z) f2-0)
+    (set! (-> out w) (sqrtf (- (- (- 1.0 (* f2-0 f2-0)) (* f1-0 f1-0)) (* f0-0 f0-0))))
+    )
+  out
+  )
+
+(defun sprite-add-matrix-data ((dma-buff dma-buffer) (matrix-mode uint))
+  "Add matrix data to a dma-buffer.
+   Mode 0 = send (-> *math-camera* camera-temp)
+   Mode 1 = screen sprites, like the HUD"
+  (let ((count 900)) ;; maybe address??
```

Comment:
yeah, I think this is the address for where the matrix goes. See `SpriteDataMem::Matrix = 900` in `sprite_common.h`.

Reply:
🙂 for what it's worth, i think these are yours or someone else's comments. i copied everything from the jak 1 code that still made sense and tweaked a little